### ISSUE TITLE
[BUG] The Installer fails to install if there are LVM vgs on the install disk

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#5398]
**Describe the bug**

The installation fails if there are existing partitions/VGs/LVs on the install disk:

```
[root@localhost ~]# lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
nvme0n1     259:0    0 894.3G  0 disk 
├─nvme0n1p1 259:1    0   600M  0 part /boot/efi
├─nvme0n1p2 259:2    0     1G  0 part /boot
└─nvme0n1p3 259:3    0 892.7G  0 part 
  ├─cl-root 253:0    0    70G  0 lvm  /
  ├─cl-swap 253:1    0  15.2G  0 lvm  [SWAP]
  └─cl-home 253:2    0 807.4G  0 lvm  /home
  ```

![Screenshot 2024-03-18 at 10 30 27 PM](https://github.com/harvester/harvester/assets/1691518/a6e4738d-65bc-48ce-a302-e9e6b5b812dd)

**To Reproduce**
1. Create VGs and LVs on a disk.
2. Install Harvester to that disk.

**Expected behavior**

The installer should clean the install disk before the installation to prevent unexpected errors.

**Support bundle**

**Environment**
 - Harvester ISO version: v1.3.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

**Comment by noahgildersleeve** [comment]
I imagine this is probably a result of #4527 / #5174. Basically if you don't use the flag the installer won't wipe the partition tables every time. We added the flag, but it's a manual thing that you need to enable when running the installer either in ISO or PXE boot.

**Comment by tserong** [comment]
I'm not sure that flag is the issue. `wipeDisks` (if set) will wipe _all_ attached disks. We shouldn't have to set that flag in order to overwrite only the disk the user has explicitly chosen to install on - that should happen automatically.

**Comment by tserong** [comment]
OK, this one is weird. I managed to reproduce it, but it requires an extremely specific set of circumstances which AFAICT just shouldn't be possible in real life. I'll try to explain...

Some time in the dim and distant past (i.e. mid 2022), we added a call to `blkdeactivate` in `/usr/sbin/harv-install` (https://github.com/harvester/harvester-installer/commit/c66449f3b4c19312984b12de37decb9a26d8d4d8). The same thing was also apparently independently added to elemental-toolkit (https://github.com/frelon/elemental-toolkit/commit/6cdbe8552db3caa2b84ced2e2e430fa8996bfa34). Arguably one or the other of these is redundant, but the point is, this command is run in order to deactivate any/all LVM devices before trying to repartition the installation disk.

But, look at the error message we've got here running `blkdeactivate --lvmoptions wholevg,retry --dmoptions force,retry --errors`:

```
Deactivating block devices:
  Volume group "cl" not found
  Cannot process volume group cl
  [LVM]: deactivating Volume Group cl...  Volume group "cl" not found
 Cannot process volume group cl
skipping
```

Why did this fail? Why can't it find the VG? The only thing I can attribute this to in v1.3 is this more recent change:

  https://github.com/harvester/os2/pull/86

That change to the Harvester OS adds `global_filter = [ "r|.*/|" ]` to `/etc/lvm/lvm.conf`, in order to disable LVM. But if LVM is disabled, the VGs shouldn't have been activated in the first place, and we'd never have hit the `blkdeactivate` failure!

In case the above isn't clear, here's some further notes:

- To try to reproduce, I originally booted the Harvester installer, switched to the second console and went to create an LVM volume on the disk I planned to install on. This failed with `Device /dev/vdc excluded by filter`. I could only actually create the LVM volume after I removed the `global_filter` line from `/etc/lvm/lvm.conf`.
- I then proceeded with the installation, but it worked fine. LVM was active, so `blkdeactivate` worked and the disk was repartitioned successfully.
- I then retried the above, first removed `global_filter`, then created the LVM volume, then I added `global_filter` back in and tried the install. This time it failed, because `blkdeactivate` couldn't find the volume to deactivate with `global_filter` in place. "Aha!" I thought, now I've got it.
- Then I created the LVM volume, and did a fresh reboot of the installer.. But, because the `global_filter` is applied to `/etc/lvm/lvm.conf` on boot, the LVM volumes never activate in the first place, so `blkdeactivate` never has any trouble, and the install succeeds.
- I also tried installing Ubuntu first as @depasseg did in https://github.com/harvester/harvester/issues/2230#issuecomment-2048846716 but again, when I boot the installer, LVM doesn't activate so there's no problem.

The only thing I can think of is that maybe there's some case where systemd's automatic LVM activation somehow races with the elemental-setup-initramfs.service - if that were possible it would explain LVM volumes being activated initially (`global_filter` wouldn't be set yet), and `blkdeactivate` failing later (after `global_filter` is set), but I haven't been able to reproduce that.

TL;DR: @bk201 I'd be interested to know if you've got more details on exactly how you managed to hit this.

**Comment by depasseg** [comment]
@tserong here's a screenshot of my error (I can't remember if I provided this earlier, but it seems to apply to your testing).  This was on a Dell r730 installing to /dev/sdj from ISO.

![Screenshot 2024-04-10 215947](https://github.com/harvester/harvester/assets/3201827/c7e9fadf-8ff6-45bf-a9c9-d5ba061a0967)

**Comment by tserong** [comment]
Alright, I've found it. I couldn't reproduce the problem running Harvester in a VM, but got it immediately when I tried running on real hardware.  Check this out, running on an old Dell system, with LVM on a couple of disks from a previous openSUSE install. Right after booting the installer ISO off a USB stick, we have this:

```
baremetal:/home/rancher # lsblk
NAME                   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0                    7:0    0 555.9M  1 loop /run/rootfsbase
sda                      8:0    0 232.9G  0 disk 
├─sda1                   8:1    0  39.2M  0 part 
├─sda2                   8:2    0  10.3G  0 part 
├─sda3                   8:3    0   163M  0 part 
└─sda4                   8:4    0 222.4G  0 part 
  ├─system-home        254:13   0 140.4G  0 lvm  
  ├─system-root        254:14   0    60G  0 lvm  
  ├─system-swap        254:15   0     2G  0 lvm  
  └─system-images      254:16   0    20G  0 lvm  
sdb                      8:16   0 476.9G  0 disk 
├─ssd-build            254:0    0    32G  0 lvm  
├─ssd-libvirt          254:1    0   128G  0 lvm  
├─ssd-sles12--0        254:2    0     8G  0 lvm  
├─ssd-sles12--1        254:3    0     8G  0 lvm  
├─ssd-sles12--2        254:4    0     8G  0 lvm  
├─ssd-sles12--3        254:5    0     8G  0 lvm  
├─ssd-sles11sp3--0     254:6    0     4G  0 lvm  
├─ssd-sles11sp3--1     254:7    0     4G  0 lvm  
├─ssd-sles11sp3--2     254:8    0     4G  0 lvm  
├─ssd-sles11sp3--3     254:9    0     4G  0 lvm  
├─ssd-shared--block--1 254:10   0     4G  0 lvm  
├─ssd-opensuse131--0   254:11   0     4G  0 lvm  
└─ssd-fast--scratch    254:12   0 260.9G  0 lvm  
sdc                      8:32   1  28.9G  0 disk 
├─sdc1                   8:33   1   5.4G  0 part /run/initramfs/live
└─sdc2                   8:34   1     4M  0 part 
```

And if I try to run `blkdeactivate`, as happens in `harv-install` to deactivate LVM prior to trying to repartition the disk:

```
baremetal:/home/rancher # blkdeactivate --lvmoptions wholevg,retry --dmoptions force,retry --errors
Deactivating block devices:
  Volume group "ssd" not found
  Cannot process volume group ssd
  [LVM]: deactivating Volume Group ssd...  Volume group "ssd" not found
 Cannot process volume group ssd
skipping
  Volume group "system" not found
  Cannot process volume group system
  [LVM]: deactivating Volume Group system...  Volume group "system" not found
 Cannot process volume group system
skipping

baremetal:/home/rancher # lsblk
NAME                   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0                    7:0    0 555.9M  1 loop /run/rootfsbase
sda                      8:0    0 232.9G  0 disk 
├─sda1                   8:1    0  39.2M  0 part 
├─sda2                   8:2    0  10.3G  0 part 
├─sda3                   8:3    0   163M  0 part 
└─sda4                   8:4    0 222.4G  0 part 
  ├─system-home        254:13   0 140.4G  0 lvm  
  ├─system-root        254:14   0    60G  0 lvm  
  ├─system-swap        254:15   0     2G  0 lvm  
  └─system-images      254:16   0    20G  0 lvm  
sdb                      8:16   0 476.9G  0 disk 
├─ssd-build            254:0    0    32G  0 lvm  
├─ssd-libvirt          254:1    0   128G  0 lvm  
├─ssd-sles12--0        254:2    0     8G  0 lvm  
├─ssd-sles12--1        254:3    0     8G  0 lvm  
├─ssd-sles12--2        254:4    0     8G  0 lvm  
├─ssd-sles12--3        254:5    0     8G  0 lvm  
├─ssd-sles11sp3--0     254:6    0     4G  0 lvm  
├─ssd-sles11sp3--1     254:7    0     4G  0 lvm  
├─ssd-sles11sp3--2     254:8    0     4G  0 lvm  
├─ssd-sles11sp3--3     254:9    0     4G  0 lvm  
├─ssd-shared--block--1 254:10   0     4G  0 lvm  
├─ssd-opensuse131--0   254:11   0     4G  0 lvm  
└─ssd-fast--scratch    254:12   0 260.9G  0 lvm  
sdc                      8:32   1  28.9G  0 disk 
├─sdc1                   8:33   1   5.4G  0 part /run/initramfs/live
└─sdc2                   8:34   1     4M  0 part 
```

The LVM volumes are all still there.

Now, have a look at some select parts of the journal:

```
baremetal:/home/rancher # journalctl | grep -i 'dracut.*init\|lvm'
Apr 18 04:56:53 localhost systemd[1]: Starting dracut initqueue hook...
Apr 18 04:56:54 localhost dracut-initqueue[796]: Scanning devices sda4 sdb  for LVM volume groups
Apr 18 04:56:54 localhost dracut-initqueue[825]: WARNING: locking_type (4) is deprecated, using --sysinit --readonly.
Apr 18 04:56:54 localhost dracut-initqueue[825]: WARNING: PV /dev/sdb in VG ssd is using an old PV header, modify the VG to update.
Apr 18 04:56:54 localhost dracut-initqueue[825]: Found volume group "ssd" using metadata type lvm2
Apr 18 04:56:54 localhost dracut-initqueue[825]: WARNING: PV /dev/sda4 in VG system is using an old PV header, modify the VG to update.
Apr 18 04:56:54 localhost dracut-initqueue[825]: Found volume group "system" using metadata type lvm2
Apr 18 04:56:54 localhost dracut-initqueue[827]: WARNING: locking_type (4) is deprecated, using --sysinit --readonly.
Apr 18 04:56:54 localhost dracut-initqueue[827]: Allowing activation with --readonly --sysinit.
Apr 18 04:56:54 localhost dracut-initqueue[827]: WARNING: PV /dev/sdb in VG ssd is using an old PV header, modify the VG to update.
Apr 18 04:56:54 localhost dracut-initqueue[827]: 13 logical volume(s) in volume group "ssd" now active
Apr 18 04:56:54 localhost dracut-initqueue[827]: Allowing activation with --readonly --sysinit.
Apr 18 04:56:54 localhost dracut-initqueue[827]: WARNING: PV /dev/sda4 in VG system is using an old PV header, modify the VG to update.
Apr 18 04:56:55 localhost dracut-initqueue[827]: 4 logical volume(s) in volume group "system" now active
Apr 18 04:56:56 localhost systemd[1]: Finished dracut initqueue hook.
Apr 18 04:57:03 rancher elemental[1224]: INFO[2024-04-18T04:57:03Z] Processing stage step 'patch lvm config (add gloabl_filter)'. ( commands: 1, files: 0, ... )
Apr 18 04:57:04 rancher systemd[1]: dracut-initqueue.service: Deactivated successfully.
Apr 18 04:57:04 rancher systemd[1]: Stopped dracut initqueue hook.
Apr 18 04:57:05 rancher systemd[1]: Listening on LVM2 poll daemon socket.
Apr 18 04:57:05 rancher systemd[1]: Starting Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling...
Apr 18 04:57:05 rancher systemd[1]: Finished Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling.
Apr 18 04:57:07 rancher systemd[1]: Condition check resulted in LVM PV QE1vmx-KM7y-93OG-zlev-PpHm-P1bY-eiodEV on /dev/sdb being skipped.
Apr 18 04:57:07 rancher systemd[1]: Created slice Slice /system/lvm2-pvscan.
Apr 18 04:57:07 rancher systemd[1]: Starting LVM event activation on device 8:16...
Apr 18 04:57:07 rancher lvm[1480]:   pvscan[1480] device 8:16 /dev/sdb excluded by filter.
Apr 18 04:57:07 rancher systemd[1]: Finished LVM event activation on device 8:16.
Apr 18 04:57:07 rancher systemd[1]: Condition check resulted in LVM PV 2E6wSS-iLro-A9HE-8Ck0-3EbO-RJpk-MbQkzi on /dev/sda4 4 being skipped.
Apr 18 04:57:07 rancher systemd[1]: Starting LVM event activation on device 8:4...
Apr 18 04:57:07 rancher lvm[1489]:   pvscan[1489] device 8:4 /dev/sda4 excluded by filter.
Apr 18 04:57:07 rancher systemd[1]: Finished LVM event activation on device 8:4.
```

Note that `dracut-initqueue` sees all those disks with LVM volumes on them and activates them. This happens _before_ the `global_filter` is added to `/etc/lvm/lvm.conf` ("elemental[1224]: INFO[2024-04-18T04:57:03Z] Processing stage step 'patch lvm config (add gloabl_filter)'."). So that's why the LVM volumes are active in the first place.

Later, we can see the lvm pvscan doing nothing ("pvscan[1489] device 8:4 /dev/sda4 excluded by filter.") but that doesn't make any difference, the volumes are already active.

When the installer gets to running `blkdeactivate`, again, the lvm invocations in there ignore all the volumes because they're excluded by the now-in-place global filter.

When I was trying to reproduce previously on a VM, the LVM activation in dracut-initqueue just didn't happen in the first place:

```
test:/home/rancher # journalctl | grep -i 'dracut.*init\|lvm'
Apr 18 02:43:52 localhost systemd[1]: Starting dracut initqueue hook...
Apr 18 02:43:54 localhost systemd[1]: Finished dracut initqueue hook.
Apr 18 02:44:00 rancher elemental[1088]: INFO[2024-04-18T02:44:00Z] Processing stage step 'patch lvm config (add gloabl_filter)'. ( commands: 1, files: 0, ... )
Apr 18 02:44:00 rancher systemd[1]: dracut-initqueue.service: Deactivated successfully.
Apr 18 02:44:00 rancher systemd[1]: Stopped dracut initqueue hook.
Apr 18 02:44:01 rancher systemd[1]: Listening on LVM2 poll daemon socket.
Apr 18 02:44:01 rancher systemd[1]: Starting Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling...
Apr 18 02:44:01 rancher systemd[1]: Finished Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling.
Apr 18 02:44:03 rancher systemd[1]: Condition check resulted in LVM PV bnXKLz-J0S2-GC2A-Wyo5-A0Oa-UjN2-5HqBCd on /dev/sda3 3 being skipped.
Apr 18 02:44:03 rancher systemd[1]: Created slice Slice /system/lvm2-pvscan.
Apr 18 02:44:03 rancher systemd[1]: Starting LVM event activation on device 8:3...
Apr 18 02:44:03 rancher lvm[1346]:   pvscan[1346] device 8:3 /dev/sda3 excluded by filter.
Apr 18 02:44:03 rancher systemd[1]: Finished LVM event activation on device 8:3.
```

I have no idea why the `dracut-initqueue` LVM activation happens on bare metal but not in a VM. Did I miss some configuration in the VM? Are there weird timing differences?

Anyway, the point is, I've reproduced it, and I can fix it by temporarily removing the `global_filter` from `lvm.conf` before the call to `blkdeactivate`.

**Comment by irishgordo** [comment]
This looks great with Harvester v1.4.0-dev-20240705
Thanks for the fix @tserong :smile: :+1: 
I'll go ahead and close this out
